File: app/models/training.py
"""AI 做 T 训练领域模型。"""

from __future__ import annotations

from datetime import datetime
from typing import Any, Dict, List, Literal, Optional

from pydantic import BaseModel, ConfigDict, Field

from app.utils.timezone import now_tz


TradeSide = Literal["buy", "sell"]
TrainingMarket = Literal["CN"]


class TrainingSessionCreate(BaseModel):
    """创建训练会话请求。"""

    symbol: str = Field(..., min_length=6, max_length=10, description="股票或 ETF 代码")
    start_date: Optional[str] = Field(default=None, description="训练开始日期，YYYY-MM-DD")
    end_date: Optional[str] = Field(default=None, description="训练结束日期，YYYY-MM-DD")
    initial_cash: float = Field(default=100000, gt=0, description="初始本金")
    total_days: int = Field(default=30, ge=1, le=120, description="训练交易日数")
    market: TrainingMarket = Field(default="CN", description="市场类型")
    note: Optional[str] = Field(default=None, description="训练备注")


class TrainingAction(BaseModel):
    """训练中的一笔模拟交易。"""

    side: TradeSide = Field(..., description="交易方向")
    quantity: int = Field(..., gt=0, description="数量")
    price: float = Field(..., gt=0, description="成交价格")
    trade_date: Optional[str] = Field(default=None, description="交易日期，YYYY-MM-DD")
    reason: Optional[str] = Field(default=None, description="交易理由")


class TrainingPosition(BaseModel):
    """当前持仓。"""

    symbol: str = Field(..., description="标的代码")
    quantity: int = Field(..., ge=0, description="持仓数量")
    avg_cost: float = Field(..., ge=0, description="持仓均价")
    market_value: float = Field(..., ge=0, description="持仓市值")
    unrealized_pnl: float = Field(..., description="未实现盈亏")


class TrainingAdvice(BaseModel):
    """规则教练建议。"""

    trend_strength: str
    volume_change: str
    t_suitability: str
    risk_level: str
    chase_risk: str
    dip_buy_suitability: str
    position_range: str
    reason: str


class TrainingSessionResponse(BaseModel):
    """训练会话摘要。"""

    model_config = ConfigDict(extra="allow")

    session_id: str
    symbol: str
    symbol_name: Optional[str] = None
    market: TrainingMarket = "CN"
    start_date: str
    end_date: str = ""
    current_step: int = 0
    total_days: int = 30
    initial_cash: float = 100000
    cash: float = 100000
    positions: List[TrainingPosition] = Field(default_factory=list)
    realized_pnl: float = 0
    unrealized_pnl: float = 0
    total_equity: float = 100000
    trade_count: int = 0
    status: Literal["active", "finished", "paused"] = "active"
    created_at: datetime = Field(default_factory=now_tz)
    updated_at: datetime = Field(default_factory=now_tz)


class TrainingSessionSummary(BaseModel):
    """璁粌存档摘要銆?"""

    model_config = ConfigDict(extra="allow")

    session_id: str
    symbol: str
    symbol_name: Optional[str] = None
    market: TrainingMarket = "CN"
    start_date: str
    end_date: str = ""
    current_step: int = 0
    total_days: int = 30
    initial_cash: float = 100000
    cash: float = 100000
    total_equity: float = 100000
    trade_count: int = 0
    status: Literal["active", "finished", "paused"] = "active"
    note: Optional[str] = None
    created_at: datetime = Field(default_factory=now_tz)
    updated_at: datetime = Field(default_factory=now_tz)


class TrainingReplayStep(BaseModel):
    """训练回放当前可见窗口。"""

    trade_date: str
    bar_index: int
    visible_bars: List[Dict[str, Any]] = Field(default_factory=list)
    is_finished: bool = False
    current_price: Optional[float] = None
    advice: Optional[TrainingAdvice] = None
    session: Optional[TrainingSessionResponse] = None


class TrainingReport(BaseModel):
    """赛后复盘报告。"""

    session_id: str
    symbol: str
    start_date: str
    end_date: str
    final_cash: float
    final_equity: float
    realized_pnl: float
    unrealized_pnl: float
    active_return: float
    buy_and_hold_return: float
    excess_return: float
    trade_count: int
    max_drawdown: float
    good_trades: List[Dict[str, Any]] = Field(default_factory=list)
    bad_trades: List[Dict[str, Any]] = Field(default_factory=list)
    advice: str

File: frontend/src/api/training.ts
import { request, type ApiResponse, type RequestConfig } from './request'

export interface TrainingSessionCreateRequest {
  symbol: string
  start_date?: string
  end_date?: string
  initial_cash?: number
  total_days?: number
  market?: 'CN'
  note?: string
}

export interface TrainingActionRequest {
  side: 'buy' | 'sell'
  quantity: number
  price: number
  trade_date?: string
  reason?: string
}

export interface TrainingPosition {
  symbol: string
  quantity: number
  avg_cost: number
  market_value: number
  unrealized_pnl: number
}

export interface TrainingAdvice {
  trend_strength: string
  volume_change: string
  t_suitability: string
  risk_level: string
  chase_risk: string
  dip_buy_suitability: string
  position_range: string
  reason: string
}

export interface TrainingSessionResponse {
  session_id: string
  symbol: string
  symbol_name?: string
  market?: 'CN'
  start_date: string
  end_date: string
  current_step: number
  total_days: number
  initial_cash: number
  cash: number
  positions: TrainingPosition[]
  realized_pnl: number
  unrealized_pnl: number
  total_equity: number
  trade_count: number
  status: 'active' | 'finished' | 'paused'
  created_at: string
  updated_at: string
}

export interface TrainingSessionSummary {
  session_id: string
  symbol: string
  symbol_name?: string
  market?: 'CN'
  start_date: string
  end_date: string
  current_step: number
  total_days: number
  initial_cash: number
  cash: number
  total_equity: number
  trade_count: number
  active_return?: number | null
  buy_and_hold_return?: number | null
  excess_return?: number | null
  score?: number | null
  status: 'active' | 'finished' | 'paused'
  note?: string | null
  created_at: string
  updated_at: string
}

export interface TrainingReplayBar {
  trade_date: string
  open?: number | null
  high?: number | null
  low?: number | null
  close?: number | null
  volume?: number | null
  amount?: number | null
}

export interface TrainingReplayStep {
  trade_date: string
  bar_index: number
  visible_bars: TrainingReplayBar[]
  is_finished: boolean
  current_price?: number | null
  advice?: TrainingAdvice | null
  session?: TrainingSessionResponse | null
}

export interface TrainingReport {
  session_id: string
  symbol: string
  start_date: string
  end_date: string
  final_cash: number
  final_equity: number
  realized_pnl: number
  unrealized_pnl: number
  active_return: number
  buy_and_hold_return: number
  excess_return: number
  score: number
  trade_count: number
  max_drawdown: number
  good_trades: Array<Record<string, any>>
  bad_trades: Array<Record<string, any>>
  advice: string
}

export const trainingApi = {
  createSession(payload: TrainingSessionCreateRequest): Promise<ApiResponse<TrainingSessionResponse>> {
    return request.post('/api/training/sessions', payload)
  },

  getSession(sessionId: string, config?: RequestConfig): Promise<ApiResponse<TrainingSessionResponse>> {
    return request.get(`/api/training/sessions/${sessionId}`, config)
  },

  listSessions(config?: RequestConfig): Promise<ApiResponse<TrainingSessionSummary[]>> {
    return request.get('/api/training/sessions', config)
  },

  getStep(sessionId: string, config?: RequestConfig): Promise<ApiResponse<TrainingReplayStep>> {
    return request.get(`/api/training/sessions/${sessionId}/step`, config)
  },

  submitAction(sessionId: string, payload: TrainingActionRequest): Promise<ApiResponse<TrainingSessionResponse>> {
    return request.post(`/api/training/sessions/${sessionId}/actions`, payload)
  },

  advance(sessionId: string): Promise<ApiResponse<TrainingReplayStep>> {
    return request.post(`/api/training/sessions/${sessionId}/advance`, {})
  },

  finish(sessionId: string): Promise<ApiResponse<TrainingReport>> {
    return request.post(`/api/training/sessions/${sessionId}/finish`, {})
  },

  getReport(sessionId: string, config?: RequestConfig): Promise<ApiResponse<TrainingReport>> {
    return request.get(`/api/training/sessions/${sessionId}/report`, config)
  },
}
